### PR TITLE
Check PETSc errors and validate BC assembly

### PR DESCRIPTION
**Pull request by masaqif**
Summary:
- Check return codes for additional PETSc/SLEPc calls in NewtonSolver, la::petsc wrappers, SLEPcEigenSolver, and fem PETSc vector helpers.
- Reject Dirichlet boundary conditions for same-mesh single bilinear form assembly when trial/test spaces differ, while preserving mixed-domain/submesh assembly.
- Let the ADIOS2 optional VTX test skip before importing VTXWriter.

Related issues: #3995, #3988, #4045.

Testing:
- git diff --check
- /opt/homebrew/bin/python3.11 -m py_compile python/dolfinx/fem/assemble.py python/dolfinx/fem/petsc.py python/test/unit/fem/test_assembler.py python/test/unit/io/test_adios2.py
- PYTHONPATH=/tmp/dolfinx-pr-ruff /opt/homebrew/bin/python3.11 -m ruff check .  (from python/)
- PYTHONPATH=/tmp/dolfinx-pr-ruff /opt/homebrew/bin/python3.11 -m ruff format --check .  (from python/)
- Not run: focused pytest, because pytest is not installed for the available Python 3.11.
- Not run: C++ build/clang tooling, because PETSc/SLEPc, clang-tidy, and clang-format are not available locally.

### CODE REVIEW

**Comment by jhale** [comment]
Thanks! Looks good, but:

PETSc errors are already being handled at https://github.com/FEniCS/dolfinx/pull/4160 where there are some remaining issues.

Please open the other two fixes as separate PRs for easier review and tracking.

File: cpp/dolfinx/nls/NewtonSolver.cpp [hunk 1]
```diff
@@ -62,7 +68,7 @@ std::pair<double, bool> converged(const nls::petsc::NewtonSolver& solver,
 void update_solution(const nls::petsc::NewtonSolver& solver, const Vec dx,
                      Vec x)
 {
-  VecAXPY(x, -solver.relaxation_parameter, dx);
+  check_petsc_error(VecAXPY(x, -solver.relaxation_parameter, dx), "VecAXPY");
```
Comment (jhale) [inline cpp/dolfinx/nls/NewtonSolver.cpp:71]: This is a new third style of checking PETSc errors which is not needed; in the other PR we've agreed to use CHECK_ERROR macro on the returned `ierr`.
